Migrate metadata generation to new SAML2-library

tvdijen opened the pull request:
This PR aims to partially migrate to the new SAML2 library. First run: metadata generation!

To do:
- [ ] Raise coverage on the `MetadataBuilder` class.
- [ ] Create a `MetadataParser` class as a replacement of the old `SAMLParser` class. This class should be able to parse XML metadata into SimpleSAMLphp's array-format. The `Signer` class should become obsolete.
- [x] Convert the ADFS module to use the new `MetadataBuilder` class. This also requires us to add a few missing classes to the [ws-security](https://github.com/tvdijen/ws-security) library and . Missing classes are:
    - [x] wst:RequestSecurityTokenResponse
    - [x] wst:RequestSecurityToken
    - [x] wsp:AppliesTo
- [ ] Ensure that metadata we generate or consume complies with: https://docs.oasis-open.org/security/saml/Post2.0/sstc-metadata-iop-os.pdf

tvdijen committed:
Create MetadataBuilder to replace the old SAMLBuilder

tvdijen committed:
Set environment variable to fix CI on pull requests

tvdijen committed:
Add clock

tvdijen committed:
Fix CI after rebasing

tvdijen committed:
Fix CI after rebasing

tvdijen committed:
Change case to match the saml2-lib

tvdijen commented:
Yes, I believe these lines are replicated one or two times, so it makes sense to refactor this into a new method!

monkeyiq committed:
Adding a new MetadataParser class using the new API

The MetadataParser is based on SAMLParser with the API being updated
to the newer saml library. This is part of the update effort to move
code that was using saml2-legacy to use saml2 instead.

Note that this is very much a WIP with only the first test
testDiscoveryResponse updated to using the new MetadataParser class.
That one test does now complete so it is a good first move toward that
whole file using the new API.

This is also rough at this stage and will need more polish once all
the tests are passing.

tvdijen committed:
Attempted fix CI

tvdijen committed:
Remove merge-file

tvdijen committed:
Fix last unit-test

tvdijen committed:
Fix coverage

tvdijen commented on tests/src/SimpleSAML/Metadata/MetadataParserTest.php:
```diff
@@ -0,0 +1,474 @@
+XML,
+        );
+
+        $entities = SAMLParser::parseDescriptorsElement($document->documentElement);
```
Shouldn't this be `MetadataParser`?

monkeyiq replied:
From the commit message "Note that this is very much a WIP with only the first test
testDiscoveryResponse updated".

tvdijen replied:
I'm very sorry, I had missed that!

monkeyiq replied:
No worries. I am still working out if I should push these intermittent changes to my branch or back to mainline along the way. My push just now has all the 10 tests converted over.

monkeyiq replied:
These 10 are now all using MetadataParser. So all should be good now :)

monkeyiq commented:
I should have the other 9 tests working today or very soon. The code is still extremely ugly, my first goal was getting the existing tests to all work on MetadataParser and then move on from there.

I pushed the starting changes up here as it was a new class and an update to the one test file. As I have point on this update I thought that wasn't going to get in the way of anything being partial.

monkeyiq committed:
All the tests in MetadataParserTest are converted now

monkeyiq commented:
My starting plan was to create MetadataParser as a copy of SAMLParser and update it to the new `SimpleSAML\SAML2\XML\md` classes. Then I would try to get the tests in `MetadataParserTest.php` to all run again. That is the place I am at currently. [ed: the 10 tests now run]

The code in MetadataParser is very rough and not "merge ready" but does what is needed for that test suite file to run and be happy. From there working out how to merge the changes in`isHiddenFromDiscoveryAV` into `isHiddenFromDiscovery` so there are not two methods and substituting out the `self::flatten` method for something more descriptive or correct.  Then making the `MetadataParser` class more of a first class citizen (fully update the code and remove redundant stuff) while still running the test suite ok is the goal.

tvdijen commented on src/SimpleSAML/Metadata/MetadataParser.php:
```diff
@@ -930,15 +998,19 @@ private static function processExtensions(mixed $element, array $parentExtension
                     $registrationInstant = $e->getRegistrationInstant();
                     if ($registrationInstant !== null) {
-                        $ret['RegistrationInfo']['instant'] = $registrationInstant;
+                        $ret['RegistrationInfo']['instant'] = $registrationInstant->toDateTime()->format("U");
```
Replace `toDateTime()->format("U")` with `getValue()` to get a properly formatted string that complies with SAML2 specifications.

monkeyiq replied:
I had to bump the unit test to expecting a string here too. 

https://github.com/simplesamlphp/simplesamlphp/pull/1901/commits/d7cc5d0a7da561027f2d7b66cafa765e12c6e727

monkeyiq committed:
Use getValue() for a string that complies with SAML2 specifications.

The unit test was also expecting an epoch integer here so that got
updated to expecting a Z time string.

monkeyiq committed:
Flatten these protocol strings here

There is a choice to either use an array of SAMLAnyURIValue for a
$SAML20Protocols function replacement and then intersect on that or we
can flatten the objects to strings an the normal array_intersect will
work ok on the same data types here.

I have code that does both things depending on which we are after.

monkeyiq committed:
cleanup

monkeyiq committed:
cleanup

monkeyiq committed:
slight method rename

monkeyiq committed:
certificate loading

monkeyiq commented:
There are still some changes in my vendor tree that I made to help things pass it seems. So working on that next...

monkeyiq committed:
Fix for the Assert() in SPSSODescriptor

monkeyiq commented:
Looking at my diffs I added some `toArray()` methods to the following. I am guessing that just adding them to `release-6.1` / `master` in saml2 would be the way to go...

It seems github doesn't want to allow me to fork both saml2 and saml2-legacy. I could do a PR from a branch in the main saml2 repo or remove my saml2-legacy and fork and PR from there but either of those seem a little messy

Files I added toArray to:
saml2/src/XML/mdui/DomainHint.php
saml2/src/XML/mdui/GeolocationHint.php
saml2/src/XML/mdui/IPHint.php
saml2/src/XML/shibmd/Scope.php

tvdijen commented:
This is strange, because you have all the admin-permissions you could possibly want on those libraries. saml2-legacy you can forget about.. That one only has the v4 version of the library.

For saml2 v6 I've created a pull request [here](https://github.com/simplesamlphp/saml2/pull/409).. Please review and let we know if you expect to need more classes to implement toArray/fromArray.

monkeyiq commented:
I now have the fork https://github.com/monkeyiq/saml2 I suspect that the saml2-legacy and saml2 repos were similar enough, or maybe even forked from each other in the past that github was not wanting to let me make a fork of both of them on my account. 

The main saml2 lib fork is more useful so I have deleted the saml2-legacy fork I had in order to make this new fork happen.

monkeyiq commented:
The new MetadataParser uses SimpleSAML\SAML2\XML\shibmd\Scope::toArray() which I had a basic version of. saml2/master has a more real version of `toArray()` in it.  I am not sure if the procedure is to 

* cherry pick things back from master to release-6.1 here 
* just copy the toArray() methods in PRs 
* something else.

I imagine I should be using saml2 6.1 branch because that is what the composer.json is calling for. It almost works against saml2/master.

tvdijen commented:
It's not really a bugfix to add those helper-methods, so I will clone master into a new release-6.2 branch and tag it

monkeyiq committed:
I am using saml2 release-6.2 now

The name of the self::flatten methods are just temporary still. They
are there to get the shape of the data into the format that the test
suite is expecting.

I will do a composer update next to make sure I have not changed
anything else locally that needs to be fed back.

monkeyiq commented:
ok so composer update seems to still have it working on the 11 tests in `./tests/src/SimpleSAML/Metadata/MetadataParserTest.php`

monkeyiq committed:
clean those old API calls away

monkeyiq committed:
no more commented out code in Test file

monkeyiq committed:
add TODO header. clean away commented use blocks

monkeyiq committed:
trim two of the flatten methods

monkeyiq committed:
possible alternate to flatten()
